### ISSUE TITLE
Implement, export, and document downloadGenomicSequencesFromUCSC()

### ISSUE THREAD

**Issue by hpages** [Bioconductor/BSgenomeForge#2]
This task is part of the "Refactor the BSgenomeForge tools" project. The goal of this project is to replace the old tools that are currently in the **BSgenome** software package with new tools that will be implemented in the **BSgenomeForge** package.

The proposed function (`downloadGenomicSequencesFromUCSC()`) is a utility function that will find and download the 2bit file from UCSC, for a given genome. It will proceed in 2 steps:
- Build the URL to the 2bit file. The 2bit file to download should always be located in `https://hgdownload.cse.ucsc.edu/goldenPath/<genome>/bigZips/`, where `<genome>` is the name of the genome (e.g. `mm10`, `hs1`, or `mpxvRivers`), and its name should always be `<genome>.2bit`. See for example the `mm10.2bit` file in https://hgdownload.cse.ucsc.edu/goldenPath/mm10/bigZips/. Do not hardcode the `http://hgdownload.cse.ucsc.edu/goldenPath` part of the URL. Instead we'll get the URL to the goldenPath via the `goldenPath.url` argument that the `downloadGenomicSequencesFromUCSC()` function will have. Like for the `getChromInfoFromUCSC()` function from the **GenomeInfoDb** package, this argument will be set to `getOption("UCSC.goldenPath.url")` by default.
- Use `download.file()` from the **utils** package to download the 2bit file.

The arguments of `downloadGenomicSequencesFromUCSC()` will be as follow:
- `genome`: A single string specifying the name of a UCSC genome.
- `goldenPath.url`: A single string specifying the URL to the UCSC goldenPath location under which the 2bit file to download is expected to be found. Default value should be `getOption("UCSC.goldenPath.url")`.
- `destdir`: A single string that specifies the path to the destination folder. Default value should be `"."`. Tilde-expansion will be supported and performed by `download.file()`.
- `method`: Will be passed down to `download.file()`.
- `quiet`: Will be passed down to `download.file()`.

We'll start with those arguments, so 5 in total. The `download.file()` function has more arguments to control the download (see `?download.file`), but, for the sake of keeping the interface as simple as possible, we won't include them all in `downloadGenomicSequencesFromUCSC`'s interface for now. However, we might add them later as the need arises.

NOTES:
- To work on this task, please fork the **BSgenomeForge** repository. Then do your work on that fork.
- Implement the function in a new file, the `downloadGenomicSequencesFromUCSC.R` file.
- Export the function by adding `export(downloadGenomicSequencesFromUCSC)` to the `NAMESPACE` file.
- Note that `download.file` is already imported (see `importFrom(utils,download.file)` in the `NAMESPACE` file).
- Add a man page for the function, with an example that downloads a _small_ 2bit file (try to find a small genome e.g. a microbial genome). Also in the example, show how to import the file in R as a DNAStringSet object.
- Don't forget to add the new files (`downloadGenomicSequencesFromUCSC.R` and `downloadGenomicSequencesFromUCSC.Rd`) to your fork before you commit your work.
- Always test your work before committing. This consists in installing the modified package, starting R, loading the package, and playing around with the new functionality. This process is called "ad hoc manual testing". Once everything behaves and looks as expected, run `R CMD build` and `R CMD check` on the package. Note that `R CMD check` should always be run on the **source tarball** produced by `R CMD build`
- Once your work is ready to be merged, please submit a PR (Pull Request).

**Comment by Simplecodez** [comment]
Hi, @hpages. I would like to work on this task. Please can you assign me?

**Comment by hpages** [comment]
I just did.

**Comment by Simplecodez** [comment]
Alright, sir. Thank you

**Comment by Simplecodez** [comment]
Good sir, i just tried implementing the downloadGenomicSequencesFromUCSC() function.
Here is the code:

```
### The function helps to download UCSC genomic sequence of a particular organism
downloadGenomicSequencesFromUCSC <- function(genome, goldenPath.url=getOption("UCSC.goldenPath.url"), destdir=".", method, quiet=FALSE){

                                 if (substr(goldenPath.url,1,7) != "http://" && substr(goldenPath.url,1,1) == "/")
                                             goldenPath.url = paste(getOption("UCSC.goldenPath.url"), goldenPath.url, sep="")
                                      
                                 else if (substr(goldenPath.url,1,7) != "http://" && substr(goldenPath.url,1,1) != "/")
                                             stop("URL must start with 'http://' or '/'") 
                                            
                                 else if (substr(goldenPath.url,1,7) == "http://" && substr(goldenPath.url,1,1) != "/") 
                                             goldenPath.url = goldenPath.url
                                 
                                 download.file(goldenPath.url, destdir, method, quiet)
                                 
                                 cat("DONE\n")
                                 }
```
I installed it but when i called the function i got this error which i don't know why

```
> downloadGenomicSequencesFromUCSC("dm3","http://hgdownload.cse.ucsc.edu/goldenPath/dm3/bigZips/dm3.2bit", "/home/emmanuel/")
DONE
Warning messages:
1: In download.file(goldenPath.url, destdir, method, quiet) :
  URL http://hgdownload.cse.ucsc.edu/goldenPath/dm3/bigZips/dm3.2bit: cannot open destfile '/home/emmanuel', reason 'Is a directory'
2: In download.file(goldenPath.url, destdir, method, quiet) :
  download had nonzero exit status

```
Please I need your insight sir.
Thank you.

**Comment by hpages** [comment]
Let me clarify what the contract of the function is.

What we want to achieve with this function is to make it _as simple as possible_ for the user to download the genomic sequences of a given UCSC genome to their computer.  So they should just be able to do:
```
downloadGenomicSequencesFromUCSC("dm3")
```
and that should work.

They shouldn't need to know the exact URL where to find the `dm3.2bit` file. If they already know that URL, it would be simpler for them to just call `download.file()` on it. They wouldn't need our `downloadGenomicSequencesFromUCSC()` function to do the job.

Furthermore, if they need to provide the URL to the 2bit file to download, then why would they also need to provide the name of the genome? The URL to the 2bit file is all the information that is needed in order to download the file. The names of the genome is not needed for that.

As explained in the description of the task at the top of this issue, what `downloadGenomicSequencesFromUCSC("dm3")` needs to do is _build_ the URL to the 2bit file. The 2bit file to download should always be located in `https://hgdownload.cse.ucsc.edu/goldenPath/<genome>/bigZips/`. Do not hardcode the `http://hgdownload.cse.ucsc.edu/goldenPath` part of this URL. Instead use whatever the `goldenPath.url` argument is set to. By default this argument is _already_ set to the correct value, which is `getOption("UCSC.goldenPath.url")`, so the user shouldn't need to specify this argument. The only reason we add this argument to the function is if the user wanted to get the 2bit file from a _mirror_ of the UCSC Genome Browser. Then they would be able to specify a different location for the goldenPath folder. But by default, the file will be downloaded from the main goldenPath repository at UCSC.

So in the end, what you need to do is much simpler than what you did. Just build the URL to the 2bit file with `paste0()`. The name of the genome (`genome`) and the URL to the goldenPath repository (`goldenPath.url`) is all you need for that. Then pass that URL to `download.file()`.

Hope this helps. Let me know how that goes.

**Comment by Simplecodez** [comment]
Okay, thank you very much sir. Let try it out again.

**Comment by Simplecodez** [comment]
Hi @hpages, I have been able to effect the corrections sir.
Below is the function and the result gotten when it is called:

```
downloadGenomicSequencesFromUCSC <- function(genome, goldenPath.url=getOption("UCSC.goldenPath.url"), destdir=".", method, quiet=FALSE)
        {
        
        if (grepl("http://hgdownload.cse.ucsc.edu/goldenPath", goldenPath.url) == TRUE)
                 goldenPath.url <- paste(goldenPath.url,"/", genome, "/bigZips","/", genome, ".2bit", sep="")
        else
        goldenPath.url <- goldenPath.url
        
        if (destdir == ".")
        {
                 destdir <- paste(destdir, genome, ".2bit", sep="") 
                 message <- paste("DONE. file was saved as ", destdir, "\n", sep="")
        }
        else 
        {
                 destdir <- destdir
                 message <- paste("DONE. file was saved as ", destdir, "\n", sep="")
        }
        cat(paste("downloading", genome, ".2bit file from", goldenPath.url, "\n", sep=" "))
        
        download.file(goldenPath.url, destdir, method, quiet)
        cat(message)
        }

```

The result gotten:
This is when the user does: downloadGenomicSequencesFromUCSC("genome") 

```
> downloadGenomicSequencesFromUCSC("sacCer1")
downloading sacCer1 .2bit file from http://hgdownload.cse.ucsc.edu/goldenPath/sacCer1/bigZips/sacCer1.2bit 
trying URL 'http://hgdownload.cse.ucsc.edu/goldenPath/sacCer1/bigZips/sacCer1.2bit'
Content type 'unknown' length 3039530 bytes (2.9 MB)
==================================================
downloaded 2.9 MB
DONE. file was saved as .sacCer1.2bit
```

This is when the user does: downloadGenomicSequencesFromUCSC("genome", destdir="Path the file is to be stored ")

```
> downloadGenomicSequencesFromUCSC("sacCer1", destdir="/home/emmanuel/sacCer1.2bit")
downloading sacCer1 .2bit file from http://hgdownload.cse.ucsc.edu/goldenPath/sacCer1/bigZips/sacCer1.2bit 
trying URL 'http://hgdownload.cse.ucsc.edu/goldenPath/sacCer1/bigZips/sacCer1.2bit'
Content type 'unknown' length 3039530 bytes (2.9 MB)
==================================================
downloaded 2.9 MB
DONE. file was saved as /home/emmanuel/sacCer1.2bit
``` 
I would love to know what you think sir.
Thank you

**Comment by hpages** [comment]
Great! That is a much better version of the function.

Let me clarify what `goldenPath.url` is about, as there seems to be some misunderstanding: If the user wants to supply the `goldenPath.url` argument, then they're expected to provide the URL to the goldenPath _folder_, not the URL to the 2bit file.

More precisely: The location of the goldenPath folder on the UCSC site is `http://hgdownload.cse.ucsc.edu/goldenPath`. This is a public repository. However, some organizations host their own private mirror of the Genome Browser. If a user working for such organization wants to download the 2bit file from this private mirror, then they will need to provide the URL to the goldenPath _folder_ that is located on that mirror. This URL would be something like `https://some.private.ucsc.genome.mirror/goldenPath/`. Note that what they are expected to provide is the URL to the goldenPath _folder_, not the URL to the 2bit file. This is because all the mirrors are expected to follow the same directory structure inside the goldenPath _folder_, that is, the relative URL to the 2bit file for sacCer1 will _always_ be `sacCer1/bigZips/sacCer1.2bit` with respect to the goldenPath _folder_. In other words, to build the URL to `sacCer1.2bit`, you just need to append `sacCer1/bigZips/sacCer1.2bit` to `goldenPath.url`. And you should do this whether the user supplied `goldenPath.url` or not. When they supply it, it's their responsibility to supply a valid URL to a goldenPath _folder_ that they have access to.

A few other things:

- Please use 4-space indentation increments in your code. Like this:
    ```
    downloadGenomicSequencesFromUCSC <- function(...)
    {
        ...
    }
    ```
- To improve code clarity, use different variable names for different things. For example, right now you're building and storing the URL to the 2bit file in the `UCSC.goldenPath.url` variable. Your code will be much clearer if you use a different variable for that, e.g. something like `file_url`. I see the same problem with `destdir`: this is supposed to contain the path to a local _directory_, not the path to the downloaded _file_. Use a different variable for that e.g. `destfile` (which BTW is the name of the `download.file` argument that will receive this value).

- Please format you `if else` statements as follow:
    ```
    if (...) {
        ...
    } else {
        ...
    }
    ```
    But do you really need this statement?

- Operations like:
    ```
    goldenPath.url <- goldenPath.url
    ```
    or
    ```
    destdir <- destdir
    ```
  don't do anything useful. More generally speaking they belong to a class of operations called _no-op_ ("no operation"), because they don't do anything. Note that there's no need to explicitly introduce this kind of operation in our code. There's no benefit, they only make our code unnecessarily longer and they hurt readability.

- We've not discussed what the function should return. Every function in R returns a value. To be clear the returned value is the value that gets stored in `res` when we do something like `res <- foo()`, it's not what the function displays or the file that it downloaded. These things are only _side effects_. In the case of `downloadGenomicSequencesFromUCSC()`, it would be useful to return the path to the downloaded file. To do this, just put `destfile` on the last line of the body of the function. alternatively you could also put `return(destfile)` but using `return()` in the last line of a function's body is not required in R. It's only required when we need to return earlier. For example:
    ```
    absolute_value <- function(x) {
        if (x >= 0)
            return(x)  # return() required
        -x  # return() not required
    }
    ```

Hopefully with that feedback you'll be able to improve your function.

They're are a few other small details that we'll need to discuss but we'll do this later.

**Comment by Simplecodez** [comment]
> Great! That is a much better version of the function.
> 
> Let me clarify what `goldenPath.url` is about, as there seems to be some misunderstanding: If the user wants to supply the `goldenPath.url` argument, then they're expected to provide the URL to the goldenPath _folder_, not the URL to the 2bit file.
> 
> More precisely: The location of the goldenPath folder on the UCSC site is `http://hgdownload.cse.ucsc.edu/goldenPath`. This is a public repository. However, some organizations host their own private mirror of the Genome Browser. If a user working for such organization wants to download the 2bit file from this private mirror, then they will need to provide the URL to the goldenPath _folder_ that is located on that mirror. This URL would be something like `https://some.private.ucsc.genome.mirror/goldenPath/`. Note that what they are expected to provide is the URL to the goldenPath _folder_, not the URL to the 2bit file. This is because all the mirrors are expected to follow the same directory structure inside the goldenPath _folder_, that is, the relative URL to the 2bit file for sacCer1 will _always_ be `sacCer1/bigZips/sacCer1.2bit` with respect to the goldenPath _folder_. In other words, to build the URL to `sacCer1.2bit`, you just need to append `sacCer1/bigZips/sacCer1.2bit` to `goldenPath.url`. And you should do this whether the user supplied `goldenPath.url` or not. When they supply it, it's their responsibility to supply a valid URL to a goldenPath _folder_ that they have access to.
> 
> A few other things:
> 
>     * Please use 4-space indentation increments in your code. Like this:
>       ```
>       downloadGenomicSequencesFromUCSC <- function(...)
>       {
>           ...
>       }
>       ```

Well noted sir.
> 
>     * To improve code clarity, use different variable names for different things. For example, right now you're building and storing the URL to the 2bit file in the `UCSC.goldenPath.url` variable. Your code will be much clearer if you use a different variable for that, e.g. something like `file_url`. I see the same problem with `destdir`: this is supposed to contain the path to a local _directory_, not the path to the downloaded _file_. Use a different variable for that e.g. `destfile` (which BTW is the name of the `download.file` argument that will receive this value).

Sorry sir, I am a bit confused about the 'destdir' argument. When I specified it as a directory eg: '/home/emmanuel/' or "." I got this error:

```
> downloadGenomicSequencesFromUCSC("sacCer1", destdir = "/home/emmanuel/") 
Warning messages:
1: In download.file(goldenPath.url, destdir, method, quiet) :
  URL http://hgdownload.cse.ucsc.edu/goldenPath/sacCer1/bigZips/sacCer1.2bit: cannot open destfile '/home/emmanuel/', reason 'Is a directory'
2: In download.file(goldenPath.url, destdir, method, quiet) :
  download had nonzero exit status

```
So i checked ?download.file and found this:

```
destfile: a character string (or vector, see the ‘url’ argument) with
          the file path where the downloaded file is to be saved.
          Tilde-expansion is performed.
```
I then changed the 'destdir' argument to /home/emmanuel/sacCer1.2bit and it downloaded.
That was the reason I tried to append the file name to the 'destdir' argument in the function.

> 
>     * Please format you `if else` statements as follow:
>       ```
>       if (...) {
>           ...
>       } else {
>           ...
>       }
>       ```

Noted sir.   
>       
>           
>             
>           
>       
>             
>           
>       
>           
>         
>       But do you really need this statement?
> 
>     * Operations like:
>       ```
>       goldenPath.url <- goldenPath.url
>       ```
>       
>       
>           
>             
>           
>       
>             
>           
>       
>           
>         
>       or
>       ```
>       destdir <- destdir
>       ```
>       
>
>           
>             
>           
>       
>             
>           
>       
>           
>         
>       don't do anything useful. More generally speaking they belong to a class of operations called _no-op_ ("no operation"), because they don't do anything. Note that there's no need to explicitly introduce this kind of operation in our code. There's no benefit, they only make our code unnecessarily longer and they hurt readability.

Alright sir, noted.
> 
>     * We've not discussed what the function should return. Every function in R returns a value. To be clear the returned value is the value that gets stored in `res` when we do something like `res <- foo()`, it's not what the function displays or the file that it downloaded. These things are only _side effects_. In the case of `downloadGenomicSequencesFromUCSC()`, it would be useful to return the path to the downloaded file. To do this, just put `destfile` on the last line of the body of the function. alternatively you could also put `return(destfile)` but using `return()` in the last line of a function's body is not required in R. It's only required when we need to return earlier. For example:
>       ```
>       absolute_value <- function(x) {
>           if (x >= 0)
>               return(x)  # return() required
>           -x  # return() not required
>       }
>       ```
> 
I will implement that sir
> 
> Hopefully with that feedback you'll be able to improve your function.
> 
> They're are a few other small details that we'll need to discuss but we'll do this later.

**Comment by Simplecodez** [comment]
I just tried implementing the corrections:
```
downloadGenomicSequencesFromUCSC <- function(genome, goldenPath.url=getOption("UCSC.goldenPath.url"), destdir=".", method, quiet)
{
    file_url <- paste(goldenPath.url,"/", genome, "/bigZips","/", genome, ".2bit", sep="")
    
    if (destdir == ".") {
        destfile <- paste(destdir, genome, ".2bit", sep="")         
    } else {
        destfile <- destdir
    }
    
    cat(paste("downloading", genome, ".2bit file from", file_url, "\n", sep=" "))
        
    download.file(file_url, destfile, method, quiet)
    
    destfile
}
```
But I didn't change download directory yet: 
```
    if (destdir == ".") {
        destfile <- paste(destdir, genome, ".2bit", sep="")         
    } else {
        destfile <- destdir
    }
```
as I am awaiting response concerning "destdir" above.
Hope everything looks as expected.
Thank you

**Comment by hpages** [comment]
> So i checked `?download.file`

As you can see in the man page, the `destfile` argument must be the path to a _file_, not to a _directory_. `destfile` must be the file path where the downloaded file is to be saved.

OTOH the `destdir` argument of the `downloadGenomicSequencesFromUCSC()` must be the path to the destination _folder_, not to the destination file.

You did this right in the case where `destdir` is `"."`: you set `destfile` to file path where the downloaded file is to be saved.

However, in the case where `destdir` is not `"."`, you're doing this:

    destfile <- destdir
    download.file(file_url, destfile, method, quiet)

This means that you're passing the path to a _directory_ to the `download.file()` function. That's not what the `download.file()` function wants. That's why when you specify `destdir="/home/emmanuel/"` in your call to `downloadGenomicSequencesFromUCSC()`, you get the error that you get (the error comes from the `download.file()` function). 

There's no reason to do something different when `destdir` is not `"."`.

**Comment by Simplecodez** [comment]
Oh, okay. I will work on that now.
Thank you

**Comment by Simplecodez** [comment]
Good day, sir.
This is the new version of the function I came up with:

```
downloadGenomicSequencesFromUCSC <- function(genome, goldenPath.url=getOption("UCSC.goldenPath.url"), destdir=".", method, quiet=FALSE)
{
    file_url <- paste(goldenPath.url,"/", genome, "/bigZips","/", genome, ".2bit", sep="")
    
    destfile <- paste(destdir, genome, ".2bit", sep="")
    
    cat(paste("downloading ", genome, " .2bit file from ", goldenPath.url, genome, "/bigZips","/", "\n", sep=""))
        
    download.file(file_url, destfile, method, quiet)
    
    destfile
} 
```

This is the result gotten from calling the function with different instances of the 'destdir' argument:

```

> downloadGenomicSequencesFromUCSC("sacCer1")
downloading sacCer1 .2bit file from http://hgdownload.cse.ucsc.edu/goldenPathsacCer1/bigZips/
trying URL 'http://hgdownload.cse.ucsc.edu/goldenPath/sacCer1/bigZips/sacCer1.2bit'
Content type 'unknown' length 3039530 bytes (2.9 MB)
==================================================
downloaded 2.9 MB

[1] ".sacCer1.2bit"

> downloadGenomicSequencesFromUCSC("sacCer1", destdir="/home/emmanuel/")
downloading sacCer1 .2bit file from http://hgdownload.cse.ucsc.edu/goldenPathsacCer1/bigZips/
trying URL 'http://hgdownload.cse.ucsc.edu/goldenPath/sacCer1/bigZips/sacCer1.2bit'
Content type 'unknown' length 3039530 bytes (2.9 MB)
==================================================
downloaded 2.9 MB

[1] "/home/emmanuel/sacCer1.2bit"

```
I would love to know what you think sir.
Thank you.

**Comment by Simplecodez** [comment]
This is the result from running ```R CMD build BSgenomeForge```

```
R CMD build BSgenomeForge
* checking for file ‘BSgenomeForge/DESCRIPTION’ ... OK
* preparing ‘BSgenomeForge’:
* checking DESCRIPTION meta-information ... OK
* checking for LF line-endings in source and make files and shell scripts
* checking for empty or unneeded directories
* building ‘BSgenomeForge_0.0.1.tar.gz’
```
This is result gotten from running ```R CMD check BSgenomeForge_0.0.1.tar.gz```

```
* using log directory ‘/home/emmanuel/BSgenomeForge.Rcheck’
* using R version 4.2.1 (2022-06-23)
* using platform: x86_64-pc-linux-gnu (64-bit)
* using session charset: UTF-8
* checking for file ‘BSgenomeForge/DESCRIPTION’ ... OK
* this is package ‘BSgenomeForge’ version ‘0.0.1’
* package encoding: UTF-8
* checking package namespace information ... OK
* checking package dependencies ... NOTE
Depends: includes the non-default packages:
  'BiocGenerics', 'S4Vectors', 'IRanges', 'GenomeInfoDb', 'Biostrings',
  'BSgenome'
Adding so many packages to the search path is excessive and importing
selectively is preferable.
* checking if this is a source package ... OK
* checking if there is a namespace ... OK
* checking for executable files ... OK
* checking for hidden files and directories ... OK
* checking for portable file names ... OK
* checking for sufficient/correct file permissions ... OK
* checking whether package ‘BSgenomeForge’ can be installed ... OK
* checking installed package size ... OK
* checking package directory ... OK
* checking DESCRIPTION meta-information ... OK
* checking top-level files ... OK
* checking for left-over files ... OK
* checking index information ... OK
* checking package subdirectories ... OK
* checking R files for non-ASCII characters ... OK
* checking R files for syntax errors ... OK
* checking whether the package can be loaded ... OK
* checking whether the package can be loaded with stated dependencies ... OK
* checking whether the package can be unloaded cleanly ... OK
* checking whether the namespace can be loaded with stated dependencies ... OK
* checking whether the namespace can be unloaded cleanly ... OK
* checking loading without being on the library search path ... OK
* checking dependencies in R code ... OK
* checking S3 generic/method consistency ... OK
* checking replacement functions ... OK
* checking foreign function calls ... OK
* checking R code for possible problems ... OK
* checking Rd files ... OK
* checking Rd metadata ... OK
* checking Rd cross-references ... OK
* checking for missing documentation entries ... OK
* checking for code/documentation mismatches ... WARNING
Functions or methods with usage in documentation object 'downloadGenomicSequencesFromUCSC' but not in code:
  ‘download.file’

* checking Rd \usage sections ... OK
* checking Rd contents ... OK
* checking for unstated dependencies in examples ... OK
* checking examples ... OK
* checking for unstated dependencies in ‘tests’ ... OK
* checking tests ... OK
  Running ‘testthat.R’
* checking PDF version of manual ... OK
* DONE
Status: 1 WARNING, 1 NOTE
```
I don't really know why I am getting this warning:
```
Functions or methods with usage in documentation object 'downloadGenomicSequencesFromUCSC' but not in code:
  ‘download.file’
```
This is the content of my ```downloadGenomicSequencesFromUCSC.Rd``` file

```
\name{downloadGenomicSequencesFromUCSC}

\alias{downloadGenomicSequencesFromUCSC}

\alias{download.file}

\title{The downloadGenomicSequencesFromUCSC function}

\description{
  A function for downloading UCSC genome .2bit sequence files
}

\usage{
## Top-level BSgenomeForge function:

downloadGenomicSequencesFromUCSC(genome, goldenPath.url=getOption("UCSC.goldenPath.url"), destdir=".", method, quiet=FALSE)

## Low-level R function

download.file(file_url, destfile, method, quiet = FALSE, mode = "w",
                   cacheOK = TRUE,
                   extra = getOption("download.file.extra"),
                   headers = NULL, ...)
}

\arguments{
  \item{genome}{
    This is the name of the UCSC genome sequence to be downloaded.
    It is used to form the download URL.
  }
  
  \item{goldenPath.url, file_url}{
   goldenPath.url: a ‘character’ string set to getOption("UCSC.goldenPath.url)
   by default.getOption("UCSC.goldenPath.url) return the goldenPath URL, 
   http://hgdownload.cse.ucsc.edu/goldenPath. This argument can be changed as needed. 
    
    file_url: a ‘character’ string (or longer vector e.g., for the
   ‘"libcurl"’ method) naming the URL of a resource to be
   downloaded. It gets it value from the goldenPath.url argument.
  }
  
  \item{destdir, destfile}{
    destdir: It is a 'character' string indicating the path where 
    the downloaded file is to be stored
    destdir is set to "." by default. The file is saved as ".(genome).2bit" 
    This file is hidden but can be seen from command line by typing 
    "ls -a" in the parent directory. 
    "cp '.(genome).2bit' 'genome.2bit'" can then be used to change 
    the name to genome.2bit which can then be accessed easily.
    
    destfile: a character string (or vector, see the ‘url’ argument) with
    the file path where the downloaded file is to be saved.
    Tilde-expansion is performed.
    
  }
  
  \item{method}{
   Method to be used for downloading files.  Current download
   methods are ‘"internal"’, ‘"wininet"’ (Windows only)
   ‘"libcurl"’, ‘"wget"’ and ‘"curl"’, and there is a value
   ‘"auto"’: see ‘Details’ and ‘Note’.

   The method can also be set through the option
   ‘"download.file.method"’: see ‘options()’.
   } 
    
  \item{quiet}{
   If ‘TRUE’, suppress status messages (if any), and the
   progress bar.
   }
  
  \item{mode}{
   character.  The mode with which to write the file.  Useful
   values are ‘"w"’, ‘"wb"’ (binary), ‘"a"’ (append) and ‘"ab"’.
   Not used for methods ‘"wget"’ and ‘"curl"’.  See also
   ‘Details’, notably about using ‘"wb"’ for Windows.
   }
   
  \item{cacheOK}{
   logical.  Is a server-side cached value acceptable?
  }
  
  \item{extra}{
   character vector of additional command-line arguments for the
   ‘"wget"’ and ‘"curl"’ methods.
  } 
   
  \item{headers}{
   named character vector of HTTP headers to use in HTTP
   requests.  It is ignored for non-HTTP URLs.  The ‘User-Agent’
   header, coming from the ‘HTTPUserAgent’ option (see
   ‘options’) is used as the first header, automatically.
  } 

  \item{...}{
   allows additional arguments to be passed, unused.
  } 
  
  }

\details{
  This function are intended for Bioconductor users who want 
  to download the genomic sequence a UCSC genome in .2bit file format.
  The function ‘downloadGenomicSequencesFromUCSC’ can be used to download 
  a single file as described by ‘goldenPath.url’ internet and store 
  it in ‘destdir’.
  
  genome is used to make the URL used in the download.file function.
  
  The ‘goldenPath.url’ must start with a scheme such as ‘http://’, 
  ‘https://’, ‘ftp://’ or ‘file://’.  Which methods support which schemes 
  varies by R version, but ‘method = "auto"’ will try to find a method
  which supports the scheme.
  
  For details on "method" and "quiet", see (?download.file).
  
  }
  
\examples{
   
   ## Download a genome.2bit file 
   ## Downloading a genome.2bit file with only the genome passed
   
   library(BSgenomeForge)
   downloadGenomicSequencesFromUCSC("sacCer1")
   
   ## Downloading a genome.2bit file with genome and destdir arguments passed
   
   library(BSgenomeForge)
   downloadGenomicSequencesFromUCSC("sacCer1", destdir="/home/emmanuel/")
   
   ## How to import the file in R as a DNAStringSet object.
   ## genome_sequence <- import("/genomePath/genome.2bit") 
    library(BSgenomeForge)  
   genome_sequence <- import("/home/emmanuel/sacCer1.2bit") 
   genome_sequence
  }
```

**Comment by hpages** [comment]
Your latest version of `downloadGenomicSequencesFromUCSC()` looks good.

A few more things before we start discussing the man page:

- There seems to be a misunderstanding about the meaning of the default value for the `destdir` argument (`"."`). Remember that this argument should always be set to the path to a _directory_ that is writable by the user. In particular, the default value (`"."`) is itself the path to a directory. On Unix-like systems and on Windows, `.` is the path to the _current_ directory. So by default, we want `downloadGenomicSequencesFromUCSC()` to download the 2bit file to the current directory.
  Also note that the downloaded file should have the same name as the remote file. In your case though, the downloaded file got renamed to `.sacCer1.2bit`. To avoid this, you want to make sure that you always insert a slash (`/`) between `destdir` and the name of the 2bit file when you form the path to the downloaded file. This will not only avoid the file renaming problem when `destdir="."` but it will also make sure that something like `downloadGenomicSequencesFromUCSC("sacCer1", destdir="/home/emmanuel")` works. Right now, it doesn't: you must use a trailing slash in `destdir` to make it work, e.g. you have to use `destdir="/home/emmanuel/"`. There's no reason that the function wouldn't work without this trailing slash.

- The standard way to actually build a _local_ path in R is with `file.path()`, rather than with `paste()`. The reason for this is that `file.path()` is portable i.e. it's guaranteed to work on any platform supported by R. This is because `file.path()` will automatically use the appropriate path separator which is `/` on  Unix-like systems and `\` on Windows. See `?file.path` for more information.

- In this call:
    ```
    paste(goldenPath.url,"/", genome, "/bigZips","/", genome, ".2bit", sep="")
    ```
  There's no reason to not pass  `"/bigZips/"` directly to the call, instead of passing the 2 strings `"/bigZips"` and `"/"` separately.
  Also let's make sure to put a space after each comma.

- Note that there are 2 alternative ways to do the above. One is with `paste0()`:
    ```
    paste0(goldenPath.url, "/", genome, "/bigZips/", genome, ".2bit")
    ```
  See `?paste0` to learn more about `paste0()`.
  And one is with `sprintf()`:
    ```
    sprintf("%s/%s/bigZips/%s.2bit", goldenPath.url, genome, genome)
    ```
  See `?sprintf` to learn more about `sprintf()`.
  Using `paste0()` or `sprintf()` is usually preferred over using `paste()`.

- About this line:
    ```
    cat(paste("downloading ", genome, " .2bit file from ", goldenPath.url, genome, "/bigZips","/", "\n", sep=""))
    ```
  Note that the `download.file()` function is already verbose by default (because `quiet` is set to `FALSE` by default), and the information that it displays about the download already provides the kind of information that is displayed by this `cat()` statement. So this `cat()` statement is not really needed: it makes our function more verbose but doesn't actually display information that is not already displayed by `download.file()`.

- No need to separate each line of code from the previous one with an empty line.

- About the returned value. It's a nice feature to have `downloadGenomicSequencesFromUCSC()` return the path to the downloaded file. This allows the user to programmatically obtain this path by with:
    ```
    filepath <- downloadGenomicSequencesFromUCSC(...)
    ```
  This will be particularly useful when `downloadGenomicSequencesFromUCSC()` is used inside a higher-level function that needs to access the downloaded file in order to process it.
  Note that a refinement that is often used in R in this kind of situation is to have the function return an _invisible value_. This is done by wrapping the returned value in `invisible()`, like this:
    ```
    downloadGenomicSequencesFromUCSC <- function(...)
    {
        ...
        invisible(destfile)
    }
    ```
  Note that this only makes a difference when we call the function without assigning its result to a variable, like here:
    ```
    downloadGenomicSequencesFromUCSC(...)
    ```
  In this case, the returned value won't be displayed. Which is ok because in this case it doesn't really need to be displayed. This is just a minor refinement to the function, in order to reduce unnecessary verbosity (a.k.a. noise) as much as possible when the function is used interactively.

**Comment by Simplecodez** [comment]
Alright sir, I will make corrections.
Thank you.

**Comment by Simplecodez** [comment]
Hi, @hpages. I have implement the corrections.
This a new version of the function:

```
downloadGenomicSequencesFromUCSC <- function(genome, goldenPath.url=getOption("UCSC.goldenPath.url"), destdir=".", method, quiet=FALSE)
{
    file_url <- paste0(goldenPath.url, "/", genome, "/bigZips/", genome, ".2bit")
    file_name <- paste0(genome, ".2bit")
    destfile <- file.path(destdir, file_name)        
    download.file(file_url, destfile, method, quiet)
    invisible(destfile)
} 
```
This is the result I got with different instances of the 'destdir' argument:

```
> downloadGenomicSequencesFromUCSC("sacCer1")
trying URL 'http://hgdownload.cse.ucsc.edu/goldenPath/sacCer1/bigZips/sacCer1.2bit'
Content type 'unknown' length 3039530 bytes (2.9 MB)
==================================================
downloaded 2.9 MB


> downloadGenomicSequencesFromUCSC("sacCer1", destdir="/home/emmanuel")
trying URL 'http://hgdownload.cse.ucsc.edu/goldenPath/sacCer1/bigZips/sacCer1.2bit'
Content type 'unknown' length 3039530 bytes (2.9 MB)
==================================================
downloaded 2.9 MB


> downloadGenomicSequencesFromUCSC("sacCer1", destdir="/home/emmanuel/")
trying URL 'http://hgdownload.cse.ucsc.edu/goldenPath/sacCer1/bigZips/sacCer1.2bit'
Content type 'unknown' length 3039530 bytes (2.9 MB)
==================================================
downloaded 2.9 MB
```
I would love to know what you think sir.
Thank you.

**Comment by hpages** [comment]
We're almost there.

A minor detail: if you place this line

    file_name <- paste0(genome, ".2bit")

before the `file_url <- paste0(...)` line, then you can use `file_name` inside `paste0()`. That way the file name is computed only once and the code is slightly simpler/cleaner. 

Two more things:

**Argument checking**

It's a good idea to perform some basic checking of the arguments supplied by the user. For example, in the unlikely situation where the user passes a `genome`, `goldenPath.url`, or `destdir` argument that is not a single string, then the function in its current form will fail with error messages that are not always easy to understand. Like here:

    downloadGenomicSequencesFromUCSC(NA)
    # trying URL 'http://hgdownload.cse.ucsc.edu/goldenPath/NA/bigZips/NA.2bit'
    # Error in download.file(file_url, destfile, method, quiet) : 
    #   cannot open URL 'http://hgdownload.cse.ucsc.edu/goldenPath/NA/bigZips/NA.2bit'
    # In addition: Warning messages:
    # 1: In download.file(file_url, destfile, method, quiet) :
    #   downloaded length 0 != reported length 196
    # 2: In download.file(file_url, destfile, method, quiet) :
    #   cannot open URL 'http://hgdownload.cse.ucsc.edu/goldenPath/NA/bigZips/NA.2bit': HTTP status was '404 Not Found'

    downloadGenomicSequencesFromUCSC(letters)
    # Error in download.file(file_url, destfile, method, quiet) : 
    #   'url' must be a length-one character vector

    downloadGenomicSequencesFromUCSC("sacCer1", destdir=letters)
    # Error in download.file(file_url, destfile, method, quiet) : 
    #   lengths of 'url' and 'destfile' must match

Even if most users won't probably try to use the function in this way, some naive users might. As you can see, the way our function handles the problems is not particularly user-friendly, in the sense that the error messages that the user gets can be a little bit cryptic.

Performing some basic argument checks will address this. This is very easy to do. See for example how the `getChromInfoFromUCSC()` function in **GenomeInfoDb** checks its `genome` and `goldenPath.url` arguments. Please do the same here for `genome`, `goldenPath.url`, and `destdir`. For `destdir`, after checking that this is a single string with `if (!isSingleString(destdir)) stop(wmsg(...))`, we can even add a check that makes sure that `destdir` actually points to an existing directory. With something like this:

    if (!dir.exists(destdir))
        stop(wmsg("'destdir' must be the path to an existing directory"))

Once you've implemented these argument checks, try to use your function in the most crazy ways and see how it behaves. Does it handle errors in a more user-friendly way?

Note that trying to break our new function is an important part of the "ad hoc manual testing" that I've mentioned before. It almost always leads to the discovery of bugs or other imperfections in our code that we can fix. 

**A little gotcha with how `download.file()` behaves**

Try:

    downloadGenomicSequencesFromUCSC("sacCer1", destdir="/home")

Did it work? Was it expected to work? Are we satisfied with how the function handled this situation? Check the man page for `download.file` to get a hint of what's going on, especially its Value section. Try to address this.

Thanks

**Comment by Simplecodez** [comment]
Lol, it really got me. Thank you so much for these observations, I am really learning a lot by fixing them. I have tried making the corrections. I hope everything looks and works as expected.
This is the new version of the function:
```
downloadGenomicSequencesFromUCSC <- function(genome, goldenPath.url=getOption("UCSC.goldenPath.url"), destdir=".", method, quiet=FALSE)
{   
    if (!isSingleString(genome)) 
        stop(wmsg("'genome' must be a single string")) 
    if (!isSingleString(goldenPath.url)) 
        stop(wmsg("'goldenPath.url' must be a single string"))
    if (!isSingleString(destdir)) 
        stop(wmsg("'destdir' must be a single string"))
    if (!dir.exists(destdir))
        stop(wmsg("'destdir' must be the path to an existing directory"))
    if (file.access(destdir, 2))
        stop(wmsg("'destdir'  must be a path with permission"))
    if (!isTRUEorFALSE(quiet)) 
        stop(wmsg("'quiet' must be TRUE or FALSE"))
    file_name <- paste0(genome, ".2bit")
    file_url <- paste0(goldenPath.url, "/", genome, "/bigZips/", file_name)
    destfile <- file.path(destdir, file_name)        
    download.file(file_url, destfile, method, quiet)
    invisible(destfile)
}

```
I also tried adding a little check for the 'quiet' argument.

These are the results obtained from different values of the arguments

```
> downloadGenomicSequencesFromUCSC(NA)
Error in downloadGenomicSequencesFromUCSC(NA) : 
  'genome' must be a single string

> downloadGenomicSequencesFromUCSC(letters)
Error in downloadGenomicSequencesFromUCSC(letters) : 
  'genome' must be a single string

> downloadGenomicSequencesFromUCSC("sacCer1", destdir=letters)
Error in downloadGenomicSequencesFromUCSC("sacCer1", destdir = letters) : 
  'destdir' must be a single string

> downloadGenomicSequencesFromUCSC("sacCer1", destdir="/home/emman")
Error in downloadGenomicSequencesFromUCSC("sacCer1", destdir = "/home/emman") : 
  'destdir' must be the path to an existing directory

```

I also tried working on cases like this: ```downloadGenomicSequencesFromUCSC("sacCer1", destdir="/home")``` using the ```file.access()``` function. I don't really know if it is ideal, I would love to know what you think. 
This is how the error displays now. I don't really know if it is informative enough.

```
> downloadGenomicSequencesFromUCSC("sacCer1", destdir="/home")
Error in downloadGenomicSequencesFromUCSC("sacCer1", destdir = "/home") : 
  'destdir' must be a path with permission
```
Thank you

**Comment by Simplecodez** [comment]
This is the results showing successful downloads:

```
> downloadGenomicSequencesFromUCSC("sacCer1", destdir="/home/emmanuel/")
trying URL 'http://hgdownload.cse.ucsc.edu/goldenPath/sacCer1/bigZips/sacCer1.2bit'
Content type 'unknown' length 3039530 bytes (2.9 MB)
==================================================
downloaded 2.9 MB

> downloadGenomicSequencesFromUCSC("sacCer1", destdir="/home/emmanuel")
trying URL 'http://hgdownload.cse.ucsc.edu/goldenPath/sacCer1/bigZips/sacCer1.2bit'
Content type 'unknown' length 3039530 bytes (2.9 MB)
==================================================
downloaded 2.9 MB

> downloadGenomicSequencesFromUCSC("sacCer1", destdir="/home/emmanuel/")
trying URL 'http://hgdownload.cse.ucsc.edu/goldenPath/sacCer1/bigZips/sacCer1.2bit'
Content type 'unknown' length 3039530 bytes (2.9 MB)
==================================================
downloaded 2.9 MB

> downloadGenomicSequencesFromUCSC("sacCer1", destdir=".")
trying URL 'http://hgdownload.cse.ucsc.edu/goldenPath/sacCer1/bigZips/sacCer1.2bit'
Content type 'unknown' length 3039530 bytes (2.9 MB)
==================================================
downloaded 2.9 MB

> downloadGenomicSequencesFromUCSC("sacCer1")
trying URL 'http://hgdownload.cse.ucsc.edu/goldenPath/sacCer1/bigZips/sacCer1.2bit'
Content type 'unknown' length 3039530 bytes (2.9 MB)
==================================================
downloaded 2.9 MB

```

**Comment by hpages** [comment]
Excellent. Function looks really good now.

I would just reformulate this:

    stop(wmsg("'destdir'  must be a path with permission"))

to use something like this instead:

    stop(wmsg("you don't have write permission to 'destdir'"))

As for your man page, please note that it should only document your function, the `downloadGenomicSequencesFromUCSC` function. The `download.file` function doesn't need to and shouldn't be documented here. `download.file` belongs to the **utils** package and is already documented there. What you can do instead is _link_ your man page to `download.file`'s man page by adding a `\seealso` section to your man page. This is usually done with something like this:
```
\seealso{
  The \code{\link[utils]{download.file}} function in the \pkg{utils} package, that
  \code{downloadGenomicSequencesFromUCSC} uses internally to download the 2bit file.
}
```
This section should be placed towards the end of the man page, right before the `\examples` section.

Also don't forget to _export_ `downloadGenomicSequencesFromUCSC` by adding the following line to the `NAMESPACE` file:

    export(downloadGenomicSequencesFromUCSC)

Put this line at the end of the file, after all the imports.

Once you are ready, run `R CMD build` and `R CMD check` again on the package then submit a PR. There are other things that we'll need to discuss about the man page but it'll be easier to do it in the Conversation section of the PR. Thanks!

**Comment by Simplecodez** [comment]
Thank you very much sir. I have made the corrections to the man page and reformulated ```stop(wmsg("'destdir'  must be a path with permission"))``` to this: ``` stop(wmsg("you don't have write permission to 'destdir'"))``` in the function and have thus created a PR. Looking forward to the discussion about the man page.
Thank you.

**Comment by hpages** [comment]
Thanks @Simplecodez. See my comments in the PR [here](https://github.com/Bioconductor/BSgenomeForge/pull/4#issuecomment-1307948265).

**Comment by Simplecodez** [comment]
Alright sir, I have made the corrections pointed out.
Thank you

**Comment by hpages** [comment]
@Simplecodez I just merged PR #4. Thanks for your hard work on this and for not giving up. You've learned a lot in the process so I think it was worth the effort. Hopefully you feel the same. Writing succinct and accurate man pages is not easy. It's an art and it takes time to learn it.

After the merge, I've cleaned `downloadGenomicSequencesFromUCSC.Rd` by removing a bunch of trailing spaces from the file, I added you as a contributor to the package, and I bumped the version. See commit b671e60b4d20b2ce4eccfdcd939d5070ef826c65

So we can close this now. Thanks again for your contribution!

**Comment by Simplecodez** [comment]
> @Simplecodez I just merged PR #4. Thanks for your hard work on this and for not giving up. You've learned a lot in the process so I think it was worth the effort. Hopefully you feel the same. Writing succinct and accurate man pages is not easy. It's an art and it takes time to learn it.
> 
> After the merge, I've cleaned `downloadGenomicSequencesFromUCSC.Rd` by removing a bunch of trailing spaces from the file, I added you as a contributor to the package, and I bumped the version. See commit [b671e60](https://github.com/Bioconductor/BSgenomeForge/commit/b671e60b4d20b2ce4eccfdcd939d5070ef826c65)
> 
> So we can close this now. Thanks again for your contribution!

I am more than happy sir, words can't describe how I feel right now. I really learnt a lot from this. I sincerely appreciate your effort and mentorship, you made this happen.
Thank you very much sir

**Comment by Simplecodez** [comment]
I will be on the look out for more additional tasks.
Thank you.

**Comment by hpages** [comment]
The Outreachy contribution period is closed now and I don't have more tasks for you @Simplecodez, sorry. I appreciate the offer though. Best regards!
